src/main/java/uk/gov/ons/ctp/response/casesvc/service/CaseService.java

```diff
@@ -167,6 +169,86 @@ public Case findCaseByIac(final String iac) throws CTPException {
     return cazes;
   }
 
+  /**
+   * Find the paginated cases for a partyId ordered by create date descending.
+   *
+   * @param partyId the partyId
+   * @param maxCasesPerSurvey the maximum number of cases per survey to return
+   * @return the cases for the partyId
+   */
+  public List<Case> findCasesByPartyIdLimitedPerSurvey(
+      final UUID partyId, boolean iac, int maxCasesPerSurvey) {
+    log.info("Entering findCasesByPartyIdLimitedPerSurvey");
+
+    List<Case> cazes =
+        limitCasesPerSurvey(
+            caseRepo.findByPartyIdOrderByCreatedDateTimeDesc(partyId), maxCasesPerSurvey);
+
+    if (iac) {
+      cazes.stream().forEach(c -> c.setIac(caseIacAuditService.findCaseIacByCasePK(c.getCasePK())));
+    }
+
+    return cazes;
+  }
+
+  /**
+   * Return only the first n instances of case per survey The query to this in the JPA is too
+   * complex and uses SQL constructs that JPQL does not support like Row Number and Over. Hence in
+   * code
+   *
+   * @param raw, the list of cases to be potentially limited
+   * @param maxCasesPerSurvey the maximum number of cases per survey to return
+   * @return the cases for the partyId
+   */
+  public List<Case> limitCasesPerSurvey(List<Case> raw, int maxCasesPerSurvey) {
+
+    log.with("maxCasesPerSurvey", maxCasesPerSurvey).info("Entering limitCasesPerSurvey");
+
+    List<Case> limitedResults = new ArrayList();
+    HashMap<UUID, Integer> surveyCounts = new HashMap<>();
+
+    raw.stream()
+        .forEach(
+            c -> tryAddCaseToLimitedResults(c, limitedResults, surveyCounts, maxCasesPerSurvey));
+
+    return limitedResults;
+  }
+
+  /**
+   * Adds a case to a result set if the number per survey is less than maximum
+   *
+   * @param current , the current Case to consider adding
+   * @param limitedResults The limited number of results
+   * @param surveyCounts A hashmap of survey ids and counts , used to keep a running count
+   * @param maxCasesPerSurvey the maximum number of cases per survey to return
+   * @return void , adds results to passed in limited as it goes
+   */
+  private void tryAddCaseToLimitedResults(
+      Case current,
+      List<Case> limitedResults,
+      HashMap<UUID, Integer> surveyCounts,
+      int maxCasesPerSurvey) {
+
+    CaseGroup currentGroup = caseGroupRepo.findById(current.getCaseGroupId());
```

Comment:
This is a very expensive transaction if we're doing it inside the `raw.forEach` loop. Is there a way we can find this information out once rather than going back to the database every time?

Reply:
Yep , bothers me too. Not sure about making it cacheable at the repo though , so that leaves a manual cache I think ?

Reply:
Implemented a local cache with a hash map